app/services/puppet_class_importer.rb

```diff
@@ -348,8 +344,14 @@ def find_or_create_puppet_class_param(klass, param_name, value)
                                    :key_type => Foreman::ImporterPuppetclass.suggest_key_type(value))
   end
 
-  def find_or_create_puppetclass(name:)
-    puppetclass = Puppetclass.unscoped.find_or_create_by!(name: name)
+  # look for Puppet class in all scopes to make sure we do not try to create a new record
+  # with a name that already exists and hit the uniqueness constraint on name
+  def find_or_create_puppetclass_for_env(klass_name, env)
+    puppetclass = Puppetclass.find_by(name: klass_name)
+    env_class = puppetclass&.environment_classes&.find_by(environment_id: env.id)
+    puppetclass ||= Puppetclass.new(name: klass_name)
+    puppetclass.environment_classes.build(environment_id: env.id) unless env_class
```

Comment:
Would something like this be a bit cleaner?
```suggestion
    puppetclass = Puppetclass.find_or_initialize_by(name: klass_name)
    puppetclass.environment_classes.find_or_initialize_by(environment_id: env.id)
```